File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.py
"""
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        # let each element be idx, sequence length
        max_sub = 0
        stack = []
        for i in range(len(nums)):
            stack.append((i, 1))
            
        while stack:
            idx, seq_len = stack.pop()
            max_sub = max(max_sub, seq_len)
            for i in range(idx+1, len(nums)):
                if nums[idx] < nums[i]:
                    stack.append((i, seq_len+1))
                    
        return max_sub            

class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        self.cache = {}
        max_sub = 0
        for i in range(len(nums)):
            max_sub = max(max_sub, self.longest_from(i, nums))
        return max_sub    
        
    def longest_from(self, idx, nums) -> int:
        if self.cache.get(idx) != None:
            return self.cache[idx]
        
        max_sub = 1
        starts = [i for i in range(idx+1, len(nums)) if nums[idx] < nums[i]]
        for s in starts:
            max_sub = max(max_sub, 1+self.longest_from(s, nums))

        self.cache[idx] = max_sub
        return max_sub
"""
class Solution:
    def lengthOfLIS(self, nums: List[int]) -> int:
        table = [0] * len(nums)
        max_sub = 1
        for i in range(len(nums)-1, -1, -1):
            max_sub_i = 1
            for j in range(i, len(nums)):
                if nums[i] < nums[j]:
                    max_sub_i = max(max_sub_i, 1+table[j])
            table[i] = max_sub_i
            
        return max(table)            

File: 0300-longest-increasing-subsequence/0300-longest-increasing-subsequence.go
/*
func lengthOfLIS(nums []int) int {
    cache := make(map[int]int, len(nums))
    
    var longestRecursive func(int, []int) int 
    longestRecursive = func(idx int, nums []int) int {
        if v, ok := cache[idx]; ok {
            return v
        }
        maxSub := 1
        for i := idx+1; i < len(nums); i++ {
            if nums[idx] < nums[i] {
                maxSub = max(maxSub, 1+longestRecursive(i, nums))
            }
        }
        cache[idx] = maxSub
        return maxSub
    }
    
    longest := 1
    for i := 0; i < len(nums); i++ {
        longest = max(longest, longestRecursive(i, nums))
    }
    return longest
}

*/
func max(a, b int) int {
    if a < b {
        return b
    }
    return a
}

func lengthOfLIS(nums []int) int {
    table := make([]int, len(nums))
    for i := len(nums)-1; i >= 0; i-- {
        table[i] = 1
        for j := i+1; j < len(nums); j++ {
            if nums[i] < nums[j] {
                table[i] = max(table[i], 1+table[j])
            }
        }
    }
    
    longest := 0
    for _, v := range table {
        longest = max(longest, v)
    }
    return longest
}